src/algorithms/postburn/PostBurn.cc

```diff
@@ -0,0 +1,155 @@
+                }
+            }
+        }
+    }
+        else{
+
+        if(correctBeamFX){
+
+            h_beam.SetPxPyPzE(incoming_hadron[0].getMomentum().x, incoming_hadron[0].getMomentum().y, incoming_hadron[0].getMomentum().z, incoming_hadron[0].getEnergy());
+            e_beam.SetPxPyPzE(incoming_lepton[0].getMomentum().x, incoming_lepton[0].getMomentum().y, incoming_lepton[0].getMomentum().z, incoming_lepton[0].getEnergy());
+
+        }
+        else{
+
+            h_beam.SetPxPyPzE(crossingAngle*incoming_hadron[0].getEnergy(), 0.0, incoming_hadron[0].getEnergy(), incoming_hadron[0].getEnergy());
+            e_beam.SetPxPyPzE(0.0, 0.0, -incoming_lepton[0].getEnergy(), incoming_lepton[0].getEnergy());
+
+        }
+    }
+
+
+
+    //Calculate boost vectors and rotations here
+
+    ROOT::Math::PxPyPzEVector cm_frame_boost = e_beam + h_beam;
+    ROOT::Math::PxPyPzEVector tmp(-cm_frame_boost.Px(), -cm_frame_boost.Py(), -cm_frame_boost.Pz(), cm_frame_boost.E());
+
+    ROOT::Math::Boost boostVector(tmp.Px()/tmp.E(), tmp.Py()/tmp.E(), tmp.Pz()/tmp.E());
+
+    //Boost to CM frame
+    e_beam = boostVector(e_beam);
+    h_beam = boostVector(h_beam);
+
+    double rotationAngleY = -1.0*TMath::ATan2(h_beam.Px(), h_beam.Pz());
+    double rotationAngleX = 1.0*TMath::ATan2(h_beam.Py(), h_beam.Pz());
+
+    ROOT::Math::RotationY rotationAboutY(rotationAngleY);
+    ROOT::Math::RotationX rotationAboutX(rotationAngleX);
+
+    e_beam = rotationAboutY(e_beam);
+    h_beam = rotationAboutY(h_beam);
+    e_beam = rotationAboutX(e_beam);
+    h_beam = rotationAboutX(h_beam);
+
+    //Boost back to proper head-on frame
+
+    ROOT::Math::PxPyPzEVector head_on_frame_boost(0., 0., cm_frame_boost.Pz(), cm_frame_boost.E());
+    ROOT::Math::Boost headOnBoostVector(head_on_frame_boost.Px()/head_on_frame_boost.E(), head_on_frame_boost.Py()/head_on_frame_boost.E(), head_on_frame_boost.Pz()/head_on_frame_boost.E());
+
+    e_beam = headOnBoostVector(e_beam);
+    h_beam = headOnBoostVector(h_beam);
+
+    //Now, loop through events and apply operations to final-state particles
+        for (const auto& p: *mcparts) {
+
+            if(p.getGeneratorStatus() == 1) { //look for final-state particles
+            ROOT::Math::PxPyPzEVector mc(p.getMomentum().x, p.getMomentum().y, p.getMomentum().z, p.getEnergy());
+
+            mc = boostVector(mc);
+            mc = rotationAboutY(mc);
+            mc = rotationAboutX(mc);
+            mc = headOnBoostVector(mc);
+
+            edm4hep::Vector3f mcMom(mc.Px(), mc.Py(), mc.Pz());
+
+            edm4hep::MutableMCParticle MCTrack(p.clone());
+            MCTrack.setMomentum(mcMom);
+
+            if(pidUseMCTruth){
+                MCTrack.setPDG(p.getPDG());
+                MCTrack.setMass(p.getMass());
+            }
+            if(!pidUseMCTruth && pidAssumePionMass){
+                MCTrack.setPDG(211);
+                MCTrack.setMass(0.13957);
+            }
+
```

Comment:
You need the particle identity (mass) to correctly apply a boost, so yes, it belongs here when you consider a reco particle input, rather than an MC one.

Reply:
At this point, the boost had already been applied.

In any case, we prefer to have modular reconstruction. If you need to replace all masses of output particles with something else, let's add a separate alogrithm.